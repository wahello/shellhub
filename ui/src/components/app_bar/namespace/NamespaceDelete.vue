<template>
  <fragment>
    <v-btn
      color="red darken-1"
      outlined
      data-test="delete-btn"
      @click="dialog = !dialog"
    >
      Delete namespace
    </v-btn>

    <v-dialog
      v-model="dialog"
      max-width="400"
    >
      <v-card data-test="namespaceDelete-dialog">
        <v-card-title class="headline grey lighten-2 text-center">
          Are you sure?
        </v-card-title>

        <v-card-text class="mt-4 mb-3 pb-1">
          <p
            v-if="active && info.nextPaymentDue !== null"
            data-test="contentSubscription-p"
          >
            Deleting the namespace will generate an invoice,
            estimated <b> {{ info.nextPaymentDue | formatCurrency }} </b> for the time of use.
          </p>

          <p data-test="content-text">
            This action cannot be undone. This will permanently delete the
            <b> {{ displayOnlyTenCharacters(name) }} </b>and its related data.
          </p>
        </v-card-text>

        <v-card-actions>
          <v-spacer />

          <v-btn
            text
            data-test="close-btn"
            @click="dialog=!dialog"
          >
            Close
          </v-btn>

          <v-btn
            color="red darken-1"
            text
            data-test="remove-btn"
            @click="remove();"
          >
            Remove
          </v-btn>
        </v-card-actions>
      </v-card>
    </v-dialog>
  </fragment>
</template>

<script>

import formatCurrency from '@/components/filter/currency';

export default {
  name: 'NamespaceDelete',

  filters: { formatCurrency },

  props: {
    nsTenant: {
      type: String,
      required: true,
    },
  },

  data() {
    return {
      name: '',
      dialog: false,
      card: null,
      elements: null,
      info: {},
      cardItems: {},
    };
  },

  computed: {
    tenant() {
      return this.$props.nsTenant;
    },

    active() {
      return this.$store.getters['billing/active'];
    },

    billing() {
      return this.$store.getters['billing/get'];
    },

    isOwner() {
      return this.$store.getters['namespaces/owner'];
    },
  },

  async created() {
    this.name = this.$store.getters['namespaces/get'].name;
  },

  mounted() {
    if (this.isOwner && !this.active) {
      this.mountStripeElements();
    }

    if (this.isOwner) {
      this.getSubscriptionInfo();
    }
  },

  methods: {
    async remove() {
      try {
        this.dialog = !this.dialog;
        await this.$store.dispatch('namespaces/remove', this.tenant);
        await this.$store.dispatch('auth/logout');
        await this.$router.push({ name: 'login' });
        this.$store.dispatch('snackbar/showSnackbarSuccessAction', this.$success.namespaceDelete);
      } catch {
        this.$store.dispatch('snackbar/showSnackbarErrorAction', this.$errors.snackbar.namespaceDelete);
      }
    },

    displayOnlyTenCharacters(str) {
      if (str !== undefined) {
        if (str.length > 10) return `${str.substr(0, 10)}...`;
      }
      return str;
    },

    mountStripeElements() {
      this.elements = this.$stripe.elements();
      this.card = this.elements.create('card');
    },

    getInfo(data) {
      const li = data.latest_invoice;
      const ui = data.upcoming_invoice;
      const description = data.product_description;
      const { card } = data;

      return {
        info: {
          periodEnd: this.billing.current_period_end,
          description,
          latestPaymentDue: li.amount_due,
          latestPaymentPaid: li.amount_paid,
          nextPaymentDue: ui.amount_due,
          nextPaymentPaid: ui.amount_paid,
        },
        card: {
          brand: card.brand,
          expYear: card.exp_year,
          expMonth: card.exp_month,
          last4: card.last4,
        },
      };
    },

    async getSubscriptionInfo() {
      if (this.active) {
        try {
          const data = await this.$store.dispatch('billing/getSubscription');
          const { info, card } = this.getInfo(data);
          this.info = info;
          this.cardItems = card;
        } catch {
          this.$store.dispatch('snackbar/showSnackbarErrorDefault');
        }
      }
    },
  },
};

</script>
