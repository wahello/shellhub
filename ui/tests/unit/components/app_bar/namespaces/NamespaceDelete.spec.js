import Vuex from 'vuex';
import { mount, createLocalVue } from '@vue/test-utils';
import Vuetify from 'vuetify';
import NamespaceDelete from '@/components/app_bar/namespace/NamespaceDelete';

describe('NamespaceDelete', () => {
  const localVue = createLocalVue();
  const vuetify = new Vuetify();
  localVue.use(Vuex);

  document.body.setAttribute('data-app', true);

  let wrapper;

  const tenant = 'xxxxxx';
  const owner = true;

  const namespace = {
    name: 'namespace3',
    owner: 'user1',
    member_names: ['user6', 'user7', 'user8'],
    tenant_id: 'a736a52b-5777-4f92-b0b8-e359bf484715',
  };

  const inactiveBilling = {
    active: false,
    current_period_end: 0,
    customer_id: '',
    subscription_id: '',
    payment_method_id: '',
  };

  const activeBilling = {
    active: true,
    current_period_end: 12121,
    customer_id: 'cus_123',
    subscription_id: 'subs_123',
    payment_method_id: 'pm_123',
  };

  const text = `This action cannot be undone. This will permanently delete the
           ${namespace.name} and its related data.`;

  const storeOwnerWithoutSubscription = new Vuex.Store({
    namespaced: true,
    state: {
      namespace,
      stateBilling: inactiveBilling.active,
      billing: inactiveBilling,
      owner,
    },
    getters: {
      'namespaces/get': (state) => state.namespace,
      'billing/active': (state) => state.stateBilling,
      'billing/get': (state) => state.billing,
      'namespaces/owner': (state) => state.owner,
    },
    actions: {
      'namespaces/remove': () => {},
      'auth/logout': () => {},
      'billing/getSubscription': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorDefault': () => {},
    },
  });

  const storeOwnerWithSubscription = new Vuex.Store({
    namespaced: true,
    state: {
      namespace,
      stateBilling: activeBilling.active,
      billing: activeBilling,
      owner,
    },
    getters: {
      'namespaces/get': (state) => state.namespace,
      'billing/active': (state) => state.stateBilling,
      'billing/get': (state) => state.billing,
      'namespaces/owner': (state) => state.owner,
    },
    actions: {
      'namespaces/remove': () => {},
      'auth/logout': () => {},
      'billing/getSubscription': () => {},
      'snackbar/showSnackbarErrorLoading': () => {},
      'snackbar/showSnackbarSuccessAction': () => {},
      'snackbar/showSnackbarErrorDefault': () => {},
    },
  });

  ///////
  // In this case, the rendering of the dialog is checked. In which
  // case with the input data it cannot take place.
  ///////

  describe('Button', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceDelete, {
        store: storeOwnerWithoutSubscription,
        localVue,
        stubs: ['fragment'],
        propsData: { nsTenant: tenant },
        vuetify,
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receives data in props', () => {
      expect(wrapper.vm.nsTenant).toEqual(tenant);
    });
    it('Compare data with the default', () => {
      expect(wrapper.vm.name).toEqual(namespace.name);
      expect(wrapper.vm.dialog).toEqual(false);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', async () => {
      expect(wrapper.find('[data-test="delete-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="namespaceDelete-dialog"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="contentSubscription-p"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="remove-btn"]').exists()).toEqual(false);
    });
  });

  ///////
  // In this case, when the user owns the keys and the focus of
  // the test is dialog rendering.
  ///////

  describe('Dialog without subscription', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceDelete, {
        store: storeOwnerWithoutSubscription,
        localVue,
        stubs: ['fragment'],
        propsData: { nsTenant: tenant },
        vuetify,
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receives data in props', () => {
      expect(wrapper.vm.nsTenant).toEqual(tenant);
    });
    it('Compare data with the default', () => {
      expect(wrapper.vm.name).toEqual(namespace.name);
      expect(wrapper.vm.dialog).toEqual(true);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="delete-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="namespaceDelete-dialog"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="contentSubscription-p"]').exists()).toEqual(false);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="remove-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="content-text"]').text()).toEqual(text);
    });
  });

  ///////
  // In this case, when the user owns the keys and the focus of
  // the test is dialog rendering.
  ///////

  describe('Dialog with subscription', () => {
    beforeEach(() => {
      wrapper = mount(NamespaceDelete, {
        store: storeOwnerWithSubscription,
        localVue,
        stubs: ['fragment'],
        propsData: { nsTenant: tenant },
        vuetify,
        mocks: {
          $stripe: {
            elements: () => ({
              create: () => ({
                mount: () => null,
              }),
            }),
          },
        },
      });

      wrapper.setData({ dialog: true });
    });

    ///////
    // Component Rendering
    //////

    it('Is a Vue instance', () => {
      expect(wrapper).toBeTruthy();
    });
    it('Renders the component', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    ///////
    // Data and Props checking
    //////

    it('Receives data in props', () => {
      expect(wrapper.vm.nsTenant).toEqual(tenant);
    });
    it('Compare data with the default', () => {
      expect(wrapper.vm.name).toEqual(namespace.name);
      expect(wrapper.vm.dialog).toEqual(true);
    });

    //////
    // HTML validation
    //////

    it('Renders the template with data', () => {
      expect(wrapper.find('[data-test="delete-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="namespaceDelete-dialog"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="contentSubscription-p"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="close-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="remove-btn"]').exists()).toEqual(true);
      expect(wrapper.find('[data-test="content-text"]').text()).toEqual(text);
    });
  });
});
