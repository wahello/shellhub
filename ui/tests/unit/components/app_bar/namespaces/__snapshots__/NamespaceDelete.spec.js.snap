// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NamespaceDelete Button Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default red--text text--darken-1" data-test="delete-btn"><span class="v-btn__content">
    Delete namespace
  </span></button>
  <div role="dialog" class="v-dialog__container">
    <!---->
  </div>
</fragment-stub>
`;

exports[`NamespaceDelete Dialog with subscription Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default red--text text--darken-1" data-test="delete-btn"><span class="v-btn__content">
    Delete namespace
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;

exports[`NamespaceDelete Dialog without subscription Renders the component 1`] = `
<fragment-stub><button type="button" class="v-btn v-btn--outlined theme--light v-size--default red--text text--darken-1" data-test="delete-btn"><span class="v-btn__content">
    Delete namespace
  </span></button>
  <div role="dialog" class="v-dialog__container"></div>
</fragment-stub>
`;
